### DISCUSSION TITLE
Experimental tab-completion in some shells (e.g. zsh) adding spaces after directory paths

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#5039] [Q&A] [answered]
It appears that the experimental bash-completions, e.g. for `jj restore`, is adding a space after each tab-completion, even if the file path so far is a non-empty directory, i.e. there is potentially a more specific completion that the user wants. This is awkward, because it requires using backspace to finish typing the path, which is something I haven't seen other autocompleters do (for any tool that has file-path arguments).

I have not tested this in other shells.

**Comment by yuja** [comment]
yeah, it's annoying. afaik, it works on fish, but not on zsh (with my confugration.)

**Reply by Adda0** [reply]
Interesting that the completions work differently for each of us. The tab-completion for paths works as intended (that is, no space for non-empty folders) for me with the dynamic jj completions (versions 0.23, 0.24) for both Zsh and Bash.

**Comment by wjv** [comment]
I'll just add here what I said on Discord before realising there's an issue filed:

* The problem occurs for me on zsh 5.9 on macOS 15.2. I have not tested other configs.
* It occurred even on a clean copy of zsh started without config (`zsh -f`), so it's unlikely to be my config.
* It occurs in both emacs and vi modes.
* The problem does **not** occur if one uses "static" completions.

(I'm assuming the problem lies with `clap` rather than jj itself.)

**Reply by wjv** [reply]
See: https://github.com/clap-rs/clap/issues/5587

**Reply by evmar** [reply]
How can I use "static" completions?  I have a wholly unconfigured stock Mac zsh and jj.

**Reply by martinvonz** [reply]
See "Standard" here: https://jj-vcs.github.io/jj/latest/install-and-setup/#zsh

**Comment by tjarvstrand** [comment]
Is there any known workaround for this? I love `jj` but this is driving me nuts.

**Reply by wjv** [reply]
Only using "standard" (that is, non-dynamic) completions, I'm afraid.

**Comment by lowking** [comment]
Same as me.
Related issues:
https://github.com/zplug/zplug/issues/602
https://github.com/jj-vcs/jj/issues/6660

**Reply by lowking** [reply]
I finally found the file and solved it. The answer has been written in the issue. I hope it will be helpful to you.

**Comment by wjv** [comment]
A note for people installing jj from Homebrew on macOS:

The brew recipe for jj automatically activates dynamic completion from a completion script installed as:
* `/opt/homebrew/share/zsh/site-functions/_jj` (on Apple Silicon Macs), or
* `/usr/local/share/zsh/site-functions/_jj` (on Intel Macs).

If you want to use static completions to avoid the extra space bug, you'll have to first delete this `_jj` _every time_ you upgrade jj  using Homebrew. (Or, of course, install jj using some other methodology.)

The dynamic completions for Homebrewed jj were added here: https://github.com/Homebrew/homebrew-core/commit/cebd3546462dcf356cbdb07022b68ee4835fdb60

**Comment by wjv** [comment, marked as answer]
As of jj 0.37.0, this appears to be fixed (?), even though the upstream clap issue is still open.

**Reply by jpgrayson** [reply]
A repair for this landed upstream in https://github.com/clap-rs/clap/pull/6203 which closed https://github.com/clap-rs/clap/issues/6178, but https://github.com/clap-rs/clap/issues/5587 remains open because it seeks a more general/comprehensive approach to the problem.

The fix first appeared in [clap 4.5.54](https://github.com/clap-rs/clap/releases/tag/v4.5.54), but is not noted in the release notes.

**Reply by wjv** [reply]
Thanks, that explains a lot! I was wondering where the stealth fix came from!

**Reply by BatmanAoD** [reply]
Haven't verified this myself (I use bash rather than zsh as my daily driver) but I'm marking it as the answer; thanks!
